### PR TITLE
Draw Lines via command

### PR DESCRIPTION
**Pull request by VanillaViking**
Hello, thanks for making and maintaining gromit-mpx.
This PR aims to implement the feature requested in https://github.com/bk138/gromit-mpx/issues/161

The original feature request also suggested opacity customization, however by the looks of it, opacity seems to be application-wide rather than individual lines? I was unsure how to implement it here. Would appreciate some guidance in that matter.

I would appreciate if you could review and provide some feedback on this code :smile:

### CODE REVIEW

**Comment by bk138** [comment]
Thanks! I'll have a look ASAP!

File: src/main.h [hunk 1]
```diff
@@ -48,6 +48,7 @@
 #define GA_ACTIVATE   gdk_atom_intern ("Gromit/activate", FALSE)
 #define GA_DEACTIVATE gdk_atom_intern ("Gromit/deactivate", FALSE)
 #define GA_TOGGLE     gdk_atom_intern ("Gromit/toggle", FALSE)
+#define GA_LINE     gdk_atom_intern ("Gromit/line", FALSE)
```
Comment (bk138) [inline src/main.h:51]: minor thing: please use same indent as other lines.

File: src/main.c [hunk 2]
```diff
@@ -982,6 +983,22 @@ int main_client (int argc, char **argv, GromitData *data)
            else
              data->clientdata = "-1"; /* default to grab all */
          }
+       else if (strcmp (arg, "-l") == 0 ||
```
Comment (bk138) [inline src/main.c:986]: Please also document in the README in the form as the other cmdline options have it.

File: src/callbacks.c [hunk 3]
```diff
@@ -502,6 +510,60 @@ void on_mainapp_selection_received (GtkWidget *widget,
...
+
+          if(data->debug) 
+	    {
+	      g_printerr("DEBUG: mainapp got line data back from client:\n");
+	      g_printerr("startX startY endX endY: %d %d %d %d\n", startX, startY, endX, endY);
+	      g_printerr("color: %s\n", hex_code);
+	      g_printerr("thickness: %d\n", thickness);
+	    }
+
+	  GdkRGBA* color = g_malloc (sizeof (GdkRGBA));
```
Comment (bk138) [inline src/callbacks.c:532]: This is never freed aka leaking memory. Please fix.

File: src/callbacks.c [hunk 4]
```diff
@@ -502,6 +510,60 @@ void on_mainapp_selection_received (GtkWidget *widget,
...
+	  if(data->debug)
+	    g_printerr("DEBUG: draw line from %d %d to %d %d\n", startX, startY, endX, endY);
+
+	  cairo_set_line_width(line_ctx->paint_ctx, thickness);
+	  cairo_move_to(line_ctx->paint_ctx, startX, startY);
+	  cairo_line_to(line_ctx->paint_ctx, endX, endY);
+	  cairo_stroke(line_ctx->paint_ctx);
+
+	  data->modified = 1;
+	  gdk_window_invalidate_rect(gtk_widget_get_window(data->win), &rect, 0);
```
Comment (bk138) [inline src/callbacks.c:562]: Could you test drawing under both X11 and Wayland?

File: src/main.c [hunk 5]
```diff
@@ -982,6 +983,22 @@ int main_client (int argc, char **argv, GromitData *data)
...
+           strcmp (arg, "--line") == 0)
+         {
+           if (argc - (i+1) == 6) /* this command must have exactly 6 params */
```
Comment (bk138) [inline src/main.c:989]: This could use some more failsafe testing, i.e. that all args are of the correct type.

**Review by bk138 — CHANGES_REQUESTED** [review]
Surprisingly small PR, good work so far! Only minor change requests.

**Comment by VanillaViking** [comment]
Thanks for the review!

I mainly use Xorg, but I briefly tested this out on wayland using Hyprland and it seems to work as expected.

As for the failsafe checks, it seems that `atoi` attempts to convert any given value to an int, or returns 0 if it can't. This would take care of most type issues, so I added checks to ensure that the coords fell within the screen area. Let me know if this is okay.

Everything else should be fixed :+1:

**Comment by bk138** [comment]
Code-wise looks good to me, I also tested under Wayland and it draws lines, but only red ones. I'm probably missing out on the correct color format so I guess adding an example to the README is indeed a good idea. Thanks for the good work so far!

File: README.md [hunk 6]
```diff
@@ -96,6 +96,9 @@ Usage:
         will undo the last drawing stroke (or "-z")
     gromit-mpx --redo
         will redo the last undone drawing stroke (or "-y")
+    gromit-mpx --line <startX> <startY> <endX> <endY> <color> <thickness>
+        will draw a straight line with characteristics specified by the arguments (or "-l")
+        eg: gromit-mpx -l 200 200 400 400 #C4A7E7 6
```
Comment (bk138) [inline README.md:101]: I guess the colour needs single quoutes around it as the stuff after '#' will be treated as a comment?

**Review by bk138 — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (VanillaViking) [reply README.md]: Ah right, fixed that up. I overlooked shell scripts!

**Comment by bk138** [comment]
Thanks for the quick reply and the contribution!

**Pull request merged**